pex/resolve/pep_691/fingerprint_service.py

```diff
@@ -0,0 +1,217 @@
+# Copyright 2022 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from __future__ import absolute_import
+
+import os
+import sqlite3
+from contextlib import closing
+from itertools import repeat
+from multiprocessing.pool import ThreadPool
+
+from pex import pex_warnings
+from pex.compatibility import cpu_count
+from pex.fetcher import URLFetcher
+from pex.resolve.pep_691.api import Client
+from pex.resolve.pep_691.model import Endpoint, Project
+from pex.resolve.resolved_requirement import Fingerprint, PartialArtifact
+from pex.resolve.resolvers import MAX_PARALLEL_DOWNLOADS
+from pex.result import Error, catch
+from pex.tracer import TRACER
+from pex.typing import TYPE_CHECKING
+from pex.variables import ENV
+
+if TYPE_CHECKING:
+    from typing import Iterable, Iterator, List, Optional, Sequence, Set, Tuple, Union
+
+    import attr  # vendor:skip
+else:
+    from pex.third_party import attr
+
+
+@attr.s(frozen=True)
+class _FingerprintedURL(object):
+    url = attr.ib()  # type: str
+    fingerprint = attr.ib()  # type: Fingerprint
+
+
+@attr.s(frozen=True)
+class FingerprintService(object):
```

Comment:
Docstring may help future readers explaining why this exists.

Reply:
It's only got 1 (real) public method and that _is_ documented so I think this is already well covered. Doubly so as soon as it gets its 1st non-test use here forthcoming.

Reply:
Fair enough -- thanks